### PR TITLE
Make ReadsSparkSink write reads in an order 100% consistent with htsjdk's SAMRecordCoordinateComparator

### LINKED ISSUE #1249: ReadSparkSink produced non-coor sorted BAM
${GATK4_JAR} MarkDuplicatesSpark -I  src/test/resources/org/broadinstitute/hellbender/tools/BQSR/CEUTrio.HiSeq.WGS.b37.ch20.1m-1m1k.NA12878.bam -O tmp.junk.bam --METRICS_FILE=tmp.metrics -parallelism 1

samtools view tmp.junk.bam | less
and you'll see 1001995 ahead of 1001645
Offending records are at [20:1001995] and [20:1001645]

### PR DESCRIPTION
**Pull request by droazen**
We were not totally consistent with htsjdk BAM ordering in our Spark bam output
previously (particularly with unmapped reads that have the position of their mates).
With this change, we are.

-Modified ReadsSparkSink to operate directly on raw, headerless SAMRecords, and use a new
 comparator for headerless SAMRecords that calls directly into htsjdk's SAMRecordCoordinateComparator

-To make this performant, added a new serializer for raw SAMRecords that uses the same
 approach as the existing SAMRecordToGATKReadAdapterSerializer in using a modified BAM
 codec as the encoder.

-Added test cases that failed in the previous implementation due to ordering issues, and
 pass now.

Resolves #1249

### CODE REVIEW

**Comment by droazen** [comment]
For @akiezun to review, please.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 1]
```diff
@@ -38,6 +45,14 @@
         return new Object[][]{
                 {testDataDir + "tools/BQSR/HiSeq.1mb.1RG.2k_lines.bam", "ReadsSparkSinkUnitTest1_ADAM"},
                 {testDataDir + "tools/BQSR/expected.HiSeq.1mb.1RG.2k_lines.alternate.recalibrated.DIQ.bam", "ReadsSparkSinkUnitTest2_ADAM"},
+
+                // This file has unmapped reads that are set to the position of their mates -- the ordering check
+                // in the tests below will fail if our ordering of these reads relative to the mapped reads
+                // is not consistent with the definition of coordinate sorting as defined in
+                // htsjdk.samtools.SAMRecordCoordinateComparator
+                //
+                // Currently disabled, as this test case doesn't pass for ADAM (we have an open ticket for this)
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: add the ticket and put the number here

File: src/main/java/org/broadinstitute/hellbender/utils/read/HeaderlessSAMRecordCoordinateComparator.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import htsjdk.samtools.SAMRecordCoordinateComparator;
+
+import java.io.Serializable;
+import java.util.Comparator;
+
+/**
+ * A comparator for headerless SAMRecords that exactly matches the ordering of the
+ * {@link htsjdk.samtools.SAMRecordCoordinateComparator}
+ */
+public class HeaderlessSAMRecordCoordinateComparator implements Comparator<SAMRecord>, Serializable {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/read/HeaderlessSAMRecordCoordinateComparator.java]: final

File: src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java [hunk 3]
```diff
@@ -7,6 +7,10 @@
 import htsjdk.samtools.*;
 import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
 
+/**
+ * Efficient serializer for SAMRecordToGATKReadAdapters that uses SAMRecordSparkCodec for encoding/decoding.
+ * Assumes that the underlying SAMRecords are headerless (and clears their header if they're not).
+ */
 public class SAMRecordToGATKReadAdapterSerializer extends Serializer<SAMRecordToGATKReadAdapter> {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java]: final

File: src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordSerializer.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import com.esotericsoftware.kryo.Serializer;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import htsjdk.samtools.*;
+
+/**
+ * Efficient serializer for SAMRecords that uses SAMRecordSparkCodec for encoding/decoding.
+ * Assumes that the SAMRecords are headerless (and clears their header if they're not).
+ */
+public class SAMRecordSerializer extends Serializer<SAMRecord> {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordSerializer.java]: final

**Comment by akiezun** [comment]
:+1: a few small edits and good to merge (and enter a ticket for ADAM reads)

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java (hunk 1 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: Done https://github.com/broadinstitute/gatk/issues/1254

File: src/main/java/org/broadinstitute/hellbender/utils/read/HeaderlessSAMRecordCoordinateComparator.java (hunk 2 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/read/HeaderlessSAMRecordCoordinateComparator.java]: Done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java (hunk 3 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java]: Done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordSerializer.java (hunk 4 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordSerializer.java]: Done

**Pull request merged**